### ISSUE TITLE
[java] "Error during type resolution" most likely because runtimeClasspath is ignored (Gradle)

### ISSUE THREAD

**Issue by huxi** [pmd/pmd#3330]
**Affects PMD Version:** 
6.35.0

**Description:**
PMD reports `Error during type resolution` even though the class reported is actually available in `runtimeClasspath`.

See this [README](https://github.com/huxi/pmd-type-resolution/blob/main/README.md) for more specific output.

**Code Sample demonstrating the issue:**

Minimal example demonstrating the issue is available at https://github.com/huxi/pmd-type-resolution

**Steps to reproduce:**
- clone https://github.com/huxi/pmd-type-resolution
- execute `./gradlew`

**Running PMD through:** *Gradle 7.0.2*

```
openjdk version "11.0.11" 2021-04-20
OpenJDK Runtime Environment AdoptOpenJDK-11.0.11+9 (build 11.0.11+9)
OpenJDK 64-Bit Server VM AdoptOpenJDK-11.0.11+9 (build 11.0.11+9, mixed mode)
```

**Comment by jsotuyod** [comment]
@huxi thanks for the sample project!

This is however a Gradle bug, as the default classpath used for typeresolution is set by the PMD plugin [here](
https://github.com/gradle/gradle/blob/67a728fe2a231de721295a685b205a57bc497009/subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/PmdPlugin.java#L149-L151):

```groovy
    taskMapping.map("classpath", () ->
        sourceSet.getOutput().plus(sourceSet.getCompileClasspath()));
    }
```

This should probably be something along the lines…

```groovy
    taskMapping.map("classpath", () ->
        sourceSet.getOutput().plus(sourceSet.getCompileClasspath())).plus(sourceSet.getRuntimeClasspath());
    }
```

so as to include both runtime-only and compile-time only dependencies.

In the meantime, you can patch this up yourself in your gradle build files by manually setting the `Pmd` task classpath accordingly.

**Comment by huxi** [comment]
Reported, thanks.

**Comment by adangel** [comment]
I'm not sure, this is a problem of the runtime classpath (I've checked maven-pmd-plugin and there we use just the compile-classpath as well).

Here's the dependency configuration:

```
def project1 = project(':project1') {
    dependencies {
        implementation 'org.apache.commons:commons-compress:1.20'
    }
}

project(':project2') {
    dependencies {
        implementation project1
    }
}
```

The missing class comes in from commons-compress. project2 depends on project1, which depends on commons-compress. It looks like that gradle doesn't resolve transitive dependencies when you have a project dependency?

The missing class is found, if the dependency is explicitly added to project2.

**Comment by huxi** [comment]
But `runtimeClasspath` is resolved correctly (i.e. including transitive dependency) by Gradle:

From the output of `./gradlew :project2:dependencies`:
```
runtimeClasspath - Runtime classpath of source set 'main'.
\--- project :project1
     \--- org.apache.commons:commons-compress:1.20

testRuntimeClasspath - Runtime classpath of source set 'test'.
\--- project :project1
     \--- org.apache.commons:commons-compress:1.20
```

**Comment by jsotuyod** [comment]
@adangel this is done on purpose, to minimize the impact of changes in transitive dependencies.

Say project A depends on B which depends on C. If you use "normal" `compile` dependencias, any change on C forces both B and A to be recompiled. This way, if C is only used internally by B **and** B doesn't change it's exposed API (ie: didn't have changes itself) Gradle can skip recompiling A altogether.

If instead of `implementation` you where to use `api`,  then the behaviour would be equivalent to Maven's `compile` dependency scope.

**Comment by big-guy** [comment]
Hey, Gradle developer here.

We attempted to make some changes here in 7.2 (https://github.com/gradle/gradle/pull/17557) to combine both runtime and classpaths, so that PMD would have access to more than just the compile classpath of the current project. 

Like @jsotuyod described, Gradle projects can split their dependencies based on API/implementation. API dependencies are exposed to consumers at runtime and compile time, while implementation dependencies are only exposed at runtime. Gradle also has "compile only" dependencies, which are dependencies only used at compile time and are never exposed to consumers.

In 7.2, we attempted to fix this by resolving the compile classpath as if it was a runtime classpath, which brings in implementation dependencies that wouldn't normally be there while also adding compile only dependencies for the current project. 

You can think of the change as...
7.1: `auxclasspath` = (implementation dependencies + compile only dependencies for the current project) + transitive api dependencies.
7.2: `auxclasspath` = (implementation dependencies + compile only dependencies for the current project) + transitive implementation dependencies.

This change caused a regression (that we can fix), but I'm thinking our approach is wrong here. There are still holes with creating auxclasspath this way. If `foo` depends on `bar` and `bar` depends on `baz` _only at compile time_, when looking at the PMD auxclasspath for `foo`, we won't see `baz`. I think this means there are still opportunities for PMD to emit a warning.

My biggest problem with this approach is that the result of this classpath is artificial because it doesn't represent a classpath that would be used for either compilation or runtime. This means that the versions of dependencies resolved for auxclasspath could be different from what was used to compile the library/application and the library/application will never see this exact classpath at runtime. 

So my question is... what do PMD rules actually expect when doing type resolution?

I think there are three options:
1. PMD should analyze the project as it would be at run time. This means that dependencies that are used at compile time only would not be available. I don't *think* this is what you want.
1. PMD should analyze the project as it would be at compile time. This means that transitive implementation dependencies would be missing. This is the current behavior.
1. PMD should analyze the project with all possible classes available--even if this pulls in implementation details about other projects. This is mostly what we tried in 7.2.

When I look at the example project above, I think the library is doing the correct thing. The implementation dependency of project1 is hidden from project2. I think the warning would make sense if a missing class was in the public API of project1, but I don't understand how project2's analysis is looking at the implementation of methods in project1.

WDYT?

I'm inclined to revert any changes to Gradle until we better understand what PMD actually needs.

**Comment by oowekyala** [comment]
@big-guy Thanks for the detailed explanation. I think ideally, PMD should analyse the project as it would be at compile-time (your option 2). Say we're in your scenario:

> If `foo` depends on `bar` and `bar` depends on `baz` *only at compile time*, when looking at the PMD auxclasspath for `foo`, we won't see `baz`.

In PMD 6, PMD loads classes to perform type resolution. Loading a class from `bar` may requiring linking against some types from `baz`. Since they're not on the classpath, PMD 6 fails with a LinkageError.

In PMD 7, no class loading is performed. Missing types from `baz` will not cause an error when loading classes from `bar`. I think these missing classes won't even be observed when analysing `foo`, as their usages should be confined to code internal to `bar` (method bodies, private methods signatures, etc) —unless I'm mistaken about the meaning of the Gradle implementation scope.

So I think PMD 7 should work fine with the current gradle strategy (your option 2), and that should be the long-term strategy. In the meantime I'm not sure what to do about PMD 6 besides living with those linkage errors. My intuition is that PMD 6 *should* work better with more classes on the classpath (so with what you've done in Gradle 7.2), though you mention it caused a regression.

**Comment by jsotuyod** [comment]
PMD 7.0.0 is now finally released, and with it, we no longer do classloading / reflection (which was pulling referenced types.

As we discussed, approach 2 (compile-time classpath) should suffice, but approach 3 (everything) if needed for backwards compatibility with PMD 6.X should not be an issue at all.
